File: leetcode-js/298-binary-tree-longest-consecutive-sequence.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var longestConsecutive = function(root) {
    return search(root, null, 0);
};

function search(root, parent, len) {
    if (!root) {
        return len;
    }

    if (parent) {
        len = root.val === parent.val + 1 ? len + 1 : 1;
    } else {
        len = 1;
    }

    return Math.max(len, search(root.left, root, len), search(root.right, root, len));
}

File: leetcode-java/298-binary-tree-longest-consecutive-sequence.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public int longestConsecutive(TreeNode root) {
        return helper(root, null, 0);
    }

    public int helper(TreeNode node, TreeNode parent, int len) {
        if (node == null) {
            return len;
        }

        if (parent != null) {
            len = node.val == parent.val + 1 ? len + 1 : 1;
        } else {
            len = 1;
        }

        int left = helper(node.left, node, len);
        int right = helper(node.right, node, len);

        return Math.max(len, Math.max(left, right));
    }
}
